### ISSUE TITLE
How to understand output to generate crystal graphs without training model

### ISSUE THREAD

**Issue by vmeschke** [txie-93/cgcnn#21]
Good morning!

I was wondering if there is an easy way to generate the crystal graphs for visualization with this code. While I would like to train a model later, I currently just want to visualize my dataset, and I'm finding it particularly hard with this code.

Thank you, and wonderful project! Everything runs smoothly with essentially no error.

Vanessa

**Comment by sgbaird** [comment]
@txie-93 I think this is a related question on our part, but we are looking to extract an intermediate layer from CGCNN. Any suggestions on how to go about this?

**Comment by txie-93** [comment]
@vmeschke and @sgbaird Sorry for my slow reply. In case you haven't solved the problem, I feel the easiest way is to return the feature vector of the intermediate layer similar to the `out`. For example, you may return both `out` and `crys_fea`.

https://github.com/txie-93/cgcnn/blob/d612a69530a72ba686fca56813657b89f2440cc5/cgcnn/model.py#L152-L166

**Comment by sgbaird** [comment]
Hi @txie-93,

Thank you for your response! By including crys_fea as a second output of model.py: CrystalGraphConvNet.forward(), I also need to update main.py train() and validate()
https://github.com/txie-93/cgcnn/blob/d612a69530a72ba686fca56813657b89f2440cc5/main.py#L251
https://github.com/txie-93/cgcnn/blob/d612a69530a72ba686fca56813657b89f2440cc5/main.py#L350
to be `output, crys_fea = model(*input_var)` to avoid errors.

However, I'm a bit stuck at this point.
Each of these lines is nested inside two `for` loops:
https://github.com/txie-93/cgcnn/blob/d612a69530a72ba686fca56813657b89f2440cc5/main.py#L173-L178
train():
https://github.com/txie-93/cgcnn/blob/d612a69530a72ba686fca56813657b89f2440cc5/main.py#L226
validate():
https://github.com/txie-93/cgcnn/blob/d612a69530a72ba686fca56813657b89f2440cc5/main.py#L325

Using Spyder and debugging, `crys_fea` has the following properties for the sample-regression data: Type: `Tensor`, Size: `torch.Size([1, 128])`.  Does the bottom level `for` statement loop through each training (or validation) crystal structure?

Sterling

P.S. I've been looking into how to extract the intermediate layers using PyTorch hooks, but haven't been able to figure it out, hence the above approach. I'm relatively new to python (bulk of my experience is in MATLAB and Mathematica).

**Comment by txie-93** [comment]
> Each of these lines is nested inside two for loops:

One iteration in the outer loop trains the model with the entire dataset. One iteration in the inner loop trains the model with one batch.

> Using Spyder and debugging, crys_fea has the following properties for the sample-regression data: Type: Tensor, Size: torch.Size([1, 128]).

I think this is because you are having a batch size of 1?

I am not exactly sure what you want. Sometime people want to visualize the learned feature vectors for each crystal. These feature vectors are usually only meaningful after the training. You may want to get all the feature vectors in the testing data?

**Comment by sgbaird** [comment]
We're interested in extracting feature vectors for all the input crystals and trying to feed these into a different model. We were thinking that the first iteration might have the most "correspondence" to the original crystal structures, but we were also considering using the last iteration (i.e. I think before applying softmax?).

Since we're not exactly sure whether we want the first, last, or an intermediate iteration's feature vectors (we will probably come to know through testing), I think what I want is to be able to extract crys_fea for each crystal (both training and test) at a given iteration. If you have a way of accessing the final iteration's feature vectors, we could start with that.

> These feature vectors are usually only meaningful after the training.

Could you elaborate? (also, when you say after the training, do you mean after all iterations of the outer loop?)

**Comment by Myles091** [comment]
Hi @txie-93 ,
      I want to visualize the learned feature vectors for each crystal.  
      I want to get all the feature vectors in the testing data. I need these for my other training.

**Comment by Sudo-Raheel** [comment]
> We're interested in extracting feature vectors for all the input crystals and trying to feed these into a different model. We were thinking that the first iteration might have the most "correspondence" to the original crystal structures, but we were also considering using the last iteration (i.e. I think before applying softmax?).
> 
> Since we're not exactly sure whether we want the first, last, or an intermediate iteration's feature vectors (we will probably come to know through testing), I think what I want is to be able to extract `crys_fea` for each crystal (both training and test) at a given iteration. If you have a way of accessing the final iteration's feature vectors, we could start with that.
> 
> > These feature vectors are usually only meaningful after the training.
> 
> Could you elaborate? (also, when you say after the training, do you mean after all iterations of the outer loop?)

Hi .... I am planning on doing the same ... were you able to accomplish the things you have mentioned... i checked out your forked repository and didn't find much... Would really appreciate any help...
Regards
